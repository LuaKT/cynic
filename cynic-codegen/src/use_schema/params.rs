#[derive(Debug)]
pub struct UseSchemaParams {
    pub schema_filename: String,
}

impl UseSchemaParams {
    fn new(schema_filename: String) -> Self {
        UseSchemaParams { schema_filename }
    }
}

impl syn::parse::Parse for UseSchemaParams {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let lit_str = input.parse::<syn::LitStr>()?;

        Ok(UseSchemaParams {
            schema_filename: lit_str.value(),
        })
    }
}
