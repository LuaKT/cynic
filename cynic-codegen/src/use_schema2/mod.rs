//mod model;
mod named_type;
mod params;
mod schema_roots;
mod subtype_markers;

pub use params::UseSchemaParams;

use proc_macro2::TokenStream;
use quote::format_ident;
use syn::parse_quote;

use crate::{
    field_type::FieldType,
    ident::{to_snake_case, Ident},
    schema::{self, SchemaLoadError},
    type_index::TypeIndex,
};

use self::{named_type::NamedType, subtype_markers::SubtypeMarkers};

pub fn use_schema(input: UseSchemaParams) -> Result<TokenStream, SchemaLoadError> {
    use quote::{quote, TokenStreamExt};

    let document = crate::schema::load_schema(input.schema_filename)?;

    let mut output = TokenStream::new();

    let root_types = schema_roots::RootTypes::from_definitions(&document.definitions);
    output.append_all(quote! {
        #root_types
    });

    let mut subtype_markers = Vec::new();
    let mut named_types = Vec::new();

    // TODO: Refactor this so it's not just one big loop
    for definition in document
        .definitions
        .into_iter()
        .filter_map(type_def_from_definition)
    {
        named_types.extend(NamedType::from_def(&definition));

        match definition {
            graphql_parser::schema::TypeDefinition::Scalar(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            graphql_parser::schema::TypeDefinition::Object(def) => {
                subtype_markers.extend(SubtypeMarkers::from_object(&def));

                let object_marker_type_name = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #object_marker_type_name;
                });

                let mut field_module_contents = Vec::new();
                for field in def.fields {
                    // TODO: wonder if we need a different Ident func for this.
                    // Technically field.name is a field, but we're only using it
                    // as a type.
                    let field_marker_type_name = Ident::for_type(&field.name);
                    let field_name_literal = proc_macro2::Literal::string(&field.name);

                    // Note: See if we can get rid of TypeIndex from this
                    // call once done.  That'd be nice.
                    let field_type =
                        FieldType::from_schema_type(&field.field_type, &TypeIndex::empty());
                    let field_type_marker = field_type.to_tokens(None, parse_quote! { super });

                    field_module_contents.push(quote! {
                        pub struct #field_marker_type_name;

                        impl ::cynic::schema::Field for #field_marker_type_name {
                            type SchemaType = #field_type_marker;

                            fn name() -> &'static str {
                                #field_name_literal
                            }
                        }

                        impl ::cynic::schema::HasField<#field_marker_type_name, #field_type_marker> for super::#object_marker_type_name {}

                        // TODO: implement HasField for all the valid conversions...
                        // assuming that's even possible - implementing the deserialize might be tricky for
                        // some of them?  Need to check what's even allowed here...
                    })

                    // TODO: Handle arguments
                }

                let field_module_name = format_ident!("{}_fields", to_snake_case(&def.name));
                output.append_all(quote! {
                    pub mod #field_module_name {
                        #(#field_module_contents)*
                    }
                });
            }
            graphql_parser::schema::TypeDefinition::Interface(def) => {
                subtype_markers.push(SubtypeMarkers::from_interface(&def));

                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: the rest of this.  Presumably we need fields & HasSubtype
            }
            graphql_parser::schema::TypeDefinition::Union(def) => {
                subtype_markers.extend(SubtypeMarkers::from_union(&def));

                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: the rest of this.  Presumably we need just HasSubtype
            }
            graphql_parser::schema::TypeDefinition::Enum(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            graphql_parser::schema::TypeDefinition::InputObject(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: Handle fields
            }
        }
    }

    output.append_all(quote! {
        #(#subtype_markers)*
        #(#named_types)*
    });

    // TODO: output defs for the built in scalars so we don't have
    // to special case them elsewhere...

    Ok(output)
}

fn type_def_from_definition(definition: schema::Definition) -> Option<schema::TypeDefinition> {
    match definition {
        graphql_parser::schema::Definition::TypeDefinition(inner) => Some(inner),
        _ => None,
    }
}
