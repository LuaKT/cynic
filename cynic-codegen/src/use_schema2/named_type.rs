use proc_macro2::TokenStream;

use crate::{idents::Ident, schema::TypeDefinition};

pub struct NamedType {
    ident: super::Ident,
}

impl NamedType {
    pub fn from_def(def: &TypeDefinition) -> Option<Self> {
        match def {
            // Note: Currently we only use the NamedType lookup for members
            // of interfaces & unions - so we specifically don't generate anything for
            // scalars, inputs or enums.
            TypeDefinition::Scalar(_) => None,
            TypeDefinition::InputObject(_) => None,
            TypeDefinition::Enum(_) => None,

            TypeDefinition::Object(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
            TypeDefinition::Interface(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
            TypeDefinition::Union(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
        }
    }
}

impl quote::ToTokens for NamedType {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let rust_name = &self.ident;
        let graphql_name = proc_macro2::Literal::string(self.ident.graphql_name());

        tokens.append_all(quote! {
            impl ::cynic::schema::NamedType for #rust_name {
                fn name() -> &'static str {
                    #graphql_name
                }
            }
        });
    }
}
