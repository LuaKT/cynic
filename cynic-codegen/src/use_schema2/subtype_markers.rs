use proc_macro2::TokenStream;

use crate::{schema, Ident};

/// Outputs `HasSubtype` implementations for the given types.
#[derive(Debug)]
pub struct SubtypeMarkers {
    pub subtype: Ident,
    pub supertypes: Vec<Ident>,
}

impl SubtypeMarkers {
    pub fn from_interface(iface: &schema::InterfaceType) -> Self {
        let ident = Ident::for_type(&iface.name);

        Self {
            subtype: ident.clone(),
            supertypes: vec![ident],
        }
    }

    pub fn from_union(def: &schema::UnionType) -> Vec<Self> {
        let supertype = Ident::for_type(&def.name);

        def.types
            .iter()
            .map(|ty| SubtypeMarkers {
                subtype: Ident::for_type(ty),
                supertypes: vec![supertype.clone()],
            })
            .collect()
    }

    pub fn from_object(obj: &schema::ObjectType) -> Option<Self> {
        if obj.implements_interfaces.is_empty() {
            return None;
        }

        Some(Self {
            subtype: Ident::for_type(&obj.name),
            supertypes: obj
                .implements_interfaces
                .iter()
                .map(Ident::for_type)
                .collect(),
        })
    }
}

impl quote::ToTokens for SubtypeMarkers {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let subtype = &self.subtype;
        let supertypes = &self.supertypes;

        tokens.append_all(quote! {
            #(
                impl ::cynic::schema::HasSubtype<#subtype> for #supertypes {}
            )*
        });
    }
}
