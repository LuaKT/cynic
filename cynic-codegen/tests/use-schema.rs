use std::{io::Write, path::PathBuf, process::Stdio};

use insta::assert_snapshot;
use rstest::rstest;

use cynic_codegen::use_schema::{use_schema, QueryDslParams};
use cynic_codegen::use_schema2::{use_schema as use_schema_2, UseSchemaParams as UseSchema2Params};

// TODO: Rename this file after running snapshots

#[rstest(schema_file => [
    "graphql.jobs.graphql",
    "books.graphql",
    "starwars.schema.graphql",
    "test_cases.graphql",
])]
fn snapshot_use_schema(schema_file: &str) {
    let schema_path = PathBuf::from("../schemas/").join(schema_file);

    let tokens = use_schema(QueryDslParams {
        schema_filename: schema_path.to_str().unwrap().to_string(),
    })
    .unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

#[rstest(schema_file_two => [
    "graphql.jobs.graphql",
    "books.graphql",
    "starwars.schema.graphql",
    "test_cases.graphql",
    "../cynic/src/bin/simple.graphql"
])]
fn snapshot_use_schema_two(schema_file_two: &str) {
    let schema_path = PathBuf::from("../schemas/").join(schema_file_two);

    let tokens = use_schema_2(UseSchema2Params {
        schema_filename: schema_path.to_str().unwrap().to_string(),
    })
    .unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

fn format_code(input: String) -> String {
    let mut cmd = std::process::Command::new("rustfmt")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .expect("failed to execute rustfmt");

    write!(cmd.stdin.as_mut().unwrap(), "{}", input).unwrap();

    std::str::from_utf8(&cmd.wait_with_output().unwrap().stdout)
        .unwrap()
        .to_owned()
}
