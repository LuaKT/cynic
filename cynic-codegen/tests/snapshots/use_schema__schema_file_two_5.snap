---
source: cynic-codegen/tests/use-schema.rs
expression: "format_code(format!(\"{}\", tokens))"

---
impl ::cynic::schema::QueryRoot for Query {}
pub struct Json {}
pub struct Query;
pub mod query_fields {
    pub struct TestStruct;
    impl ::cynic::schema::Field for TestStruct {
        type SchemaType = Option<super::TestStruct>;
        fn name() -> &'static str {
            "testStruct"
        }
    }
    impl ::cynic::schema::HasField<TestStruct, Option<super::TestStruct>> for super::Query {}
    pub struct MyUnion;
    impl ::cynic::schema::Field for MyUnion {
        type SchemaType = Option<super::MyUnionType>;
        fn name() -> &'static str {
            "myUnion"
        }
    }
    impl ::cynic::schema::HasField<MyUnion, Option<super::MyUnionType>> for super::Query {}
}
pub struct TestStruct;
pub mod test_struct_fields {
    pub struct FieldOne;
    impl ::cynic::schema::Field for FieldOne {
        type SchemaType = String;
        fn name() -> &'static str {
            "fieldOne"
        }
    }
    impl ::cynic::schema::HasField<FieldOne, String> for super::TestStruct {}
    pub struct TastyCakes;
    impl ::cynic::schema::Field for TastyCakes {
        type SchemaType = super::Dessert;
        fn name() -> &'static str {
            "tastyCakes"
        }
    }
    impl ::cynic::schema::HasField<TastyCakes, super::Dessert> for super::TestStruct {}
    pub struct FieldWithInput;
    impl ::cynic::schema::Field for FieldWithInput {
        type SchemaType = super::Dessert;
        fn name() -> &'static str {
            "fieldWithInput"
        }
    }
    impl ::cynic::schema::HasField<FieldWithInput, super::Dessert> for super::TestStruct {}
    pub struct Nested;
    impl ::cynic::schema::Field for Nested {
        type SchemaType = super::Nested;
        fn name() -> &'static str {
            "nested"
        }
    }
    impl ::cynic::schema::HasField<Nested, super::Nested> for super::TestStruct {}
    pub struct OptNested;
    impl ::cynic::schema::Field for OptNested {
        type SchemaType = Option<super::Nested>;
        fn name() -> &'static str {
            "optNested"
        }
    }
    impl ::cynic::schema::HasField<OptNested, Option<super::Nested>> for super::TestStruct {}
    pub struct Dessert;
    impl ::cynic::schema::Field for Dessert {
        type SchemaType = Option<super::Dessert>;
        fn name() -> &'static str {
            "dessert"
        }
    }
    impl ::cynic::schema::HasField<Dessert, Option<super::Dessert>> for super::TestStruct {}
    pub struct Json;
    impl ::cynic::schema::Field for Json {
        type SchemaType = Option<super::Json>;
        fn name() -> &'static str {
            "json"
        }
    }
    impl ::cynic::schema::HasField<Json, Option<super::Json>> for super::TestStruct {}
}
pub struct MyUnionType {}
pub struct Nested;
pub mod nested_fields {
    pub struct AString;
    impl ::cynic::schema::Field for AString {
        type SchemaType = String;
        fn name() -> &'static str {
            "aString"
        }
    }
    impl ::cynic::schema::HasField<AString, String> for super::Nested {}
    pub struct OptString;
    impl ::cynic::schema::Field for OptString {
        type SchemaType = Option<String>;
        fn name() -> &'static str {
            "optString"
        }
    }
    impl ::cynic::schema::HasField<OptString, Option<String>> for super::Nested {}
}
pub struct Dessert {}
pub struct AnInputType {}

